### DISCUSSION TITLE
How to use scheduler object inside fastapi route?

### DISCUSSION THREAD

**Discussion by vadympop** [agronholm/apscheduler#830] [Q&A]
I want to add schedules from routes. How I can make this?

**Comment by vadympop** [comment]
@agronholm can you help me with this question?

**Comment by agronholm** [comment]
Where are you starting the scheduler? What have you tried to expose the scheduler to the application?

**Comment by jcphlux** [comment]
This is how I have set this up is by adding it to the app.state inside the middleware.
```python
    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] == "lifespan":
            async with self.scheduler:
                await self.scheduler.start_in_background()
                self.app.state.scheduler = self.scheduler  # <- add scheduler to app state.
                await self.app(scope, receive, send)
        else:
            await self.app(scope, receive, send)
```
You can get the scheduler from the request state now.
```python
def get_scheduler_from_request(request: Request) -> AsyncScheduler:
    return request.state.scheduler
```

**Reply by jcphlux** [reply]
to get it in a route you can use dependency injection.
```python
@router.post("/something")
async def something(scheduler: Annotated[AsyncScheduler, Depends(get_scheduler_from_request)]):
    # do something with scheduler

**Reply by vadympop** [reply]
Thank you for answer!

**Reply by jcphlux** [reply]
Glad I could help!!

**Comment by jcphlux** [comment]
Just a follow up for anyone needing this.
The current solution does not work if you have a lifespan function defined ex. `app = FastAPI(lifespan=lifespan)`

To resolve this I have a simplified solution that is very clean.

**main.py**
```python
from apscheduler import AsyncScheduler
from fastapi import FastAPI
from fastapi.concurrency import asynccontextmanager


@asynccontextmanager
async def lifespan(app: FastAPI):
    # Lifespan startup actions
    scheduler = AsyncScheduler()

    async with scheduler:
        await scheduler.start_in_background()
        app.state.scheduler = scheduler

        yield # Lifespan yeilds to the main app
    
    # Lifespan shutdown actions


app = FastAPI(lifespan=lifespan)
```

**depend_scheduler.py**
```python
from apscheduler import AsyncScheduler
from fastapi import Request


def get_scheduler(request: Request) -> AsyncScheduler:
    return request.app.state.scheduler
```

Then in a route you can inject the dependance. 
```python
from typing import Annotated

from apscheduler import AsyncScheduler
from fastapi import APIRouter, Depends

from depend_scheduler import get_scheduler

router = APIRouter()


@router.post("/scheduler/")
async def get_scheduler(
    scheduler: Annotated[AsyncScheduler, Depends(get_scheduler)],
):
    jobs = await scheduler.get_jobs()
```

**Reply by agronholm** [reply]
So, wait, that shuts down the scheduler and only _then_ yields? Are you sure that's what you want?

**Reply by jcphlux** [reply]
I don't see where you are seeing it being shut down. Did I miss understand how `start_in_background` works?

**Reply by jcphlux** [reply]
wait I think I see what you are talking about. Because of the context manager it will close it.

**Reply by jcphlux** [reply]
ok I have corrected it. I think it should work now. I am not logged into work right now so I cant test it.

**Reply by jcphlux** [reply]
@agronholm  looking at my edit. Is the `await scheduler.stop()` redundant as it should be called when the context manager closes?

**Reply by agronholm** [reply]
Yes, that's why it's a context manager.

**Reply by jcphlux** [reply]
ok I have it correct to reflect that. should be final now.

**Reply by jcphlux** [reply]
I found another bug that I have corrected in `get_scheduler ` you need `return request.app.state.scheduler` not r`eturn request.state.scheduler`
